lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -303,17 +303,37 @@ class ModuleEmitter : public VerilogEmitterBase,
   void emitOperation(Operation *op);
 
   void collectNamesEmitDecls(Block &block);
-  StringRef addName(Value value, StringRef name);
+
+  template <typename ValueType>
+  using NameMap =
+      llvm::DenseMap<ValueType, llvm::StringMapEntry<llvm::NoneType> *>;
+
+  template <typename ValueType>
+  StringRef addNameImpl(ValueType value, StringRef name,
```

Comment:
I'm not sure if this is the best way to do this, but I wanted to generalize this same logic to store names not just for a `Value`, but also for a `Operation *`. I set up a separate table of `Operation *` as mentioned in #407, but I want to re-use this logic for legalizing/uniquing names.

Reply:
The problem with a separate table is that we can't have Value names conflict with instance names. I personally think that one table ('identifiers`) would be best, but we run into the problem of what to key it on. When I last tried to key it on a pair (Value, Operation*) (or something like that) it came out ugly.

@lattner What do you think?

Reply:
This implementation is supposed to use the same `usedNames` set for both the `Value` and `Operation *` table. I will try to add a test that demonstrates this.

Reply:
@teqdruid please take a look at the most recent commit. Is that the case you were concerned about, and is that handling satisfactory?

Reply:
Yes, it is

Reply:
Yeah, I didn't realize there was a `usedName` set... I'm not really the guy who should be reviewing this PR. I think Chris is the only (other) person who's been in that code a fair amount. That said, this use of templates seems reasonable enough to me.

Reply:
:+1: thanks for taking a look.

Reply:
Ok, given that we want to de-dupe instances and other values like wires all together, I decided to keep a single `nameTable`, using a  `PointerUnion<Value, Operation *>` as key. This pretty much just dropped in, except for a couple places where concrete operation types were passed to function calls. Since these can be cast to either `Value` or `Operation *`, I think this caused some ambiguity. I resolved this using an explicit constructor call in those cases, which apparently handles the ambiguity.

Otherwise, the logic and test cases are the same as the previous revision.

Reply:
The pointerunion approach LGTM